Add "unik describe-compiler" CLI

miha-plesko opened the pull request:
Currently user has to consult README to see how her
application must be configured (e.g. what config files
should look like) in order to run it with UniK.
Not any more! Introduced command
```
unik describe-compiler --base <> --language <> --provider <>
```
that prints usage information to the console. 

To do this properly, I had to quickly refactor OSv compiler.

miha-plesko commented:
Hi @ilackarms ,
as we discussed yesterday, I'm sending PR with implementation of compiler self description.

Turns out that refactoring\* of the OSv compiler was needed to support this functionality. I went step further and refactored it so that it is now much easier to understand how the compiler works. This way it is easier to understand the difference between (former) osv-qemu.go and (former) osv-java.go.

Please let me know your opinion.

@gberginc - I kindly ask for a review from your side as well

miha-plesko commented:
Another thing that I would like to discuss is an upgrade of this PR. We implement function to **describe** usage of specific compiler. But we could probably be even more user-friendly and actually **prepare** project directory for specific compiler. If user would `cd` to her project dir and run:
```
$ unik prepare-compiler --base osv --language java --provider openstack
```
Then UniK could prepare `manifest.yaml` for her with all the appropriate keys. For example:
```yaml
# manifest.yaml

main_file: FILENAME
build_command: mvn package
```
Additionally, such autogenerated yaml could contain yaml comments telling user what is the meaning of each key (e.g. for `main_file` the yaml comment could explain that you must input path to your nodejs entrypojnt file there).

Thoughts?

ilackarms commented:
how will prepare-compiler work? can you take me through the workflow?

ilackarms commented:
can you explain the purpose of the `Bootstrap()` method that has been added to all the compilers?
(such as https://github.com/emc-advanced-dev/unik/pull/110/files#diff-92d0b88f7d0c433cb6ff45d968157a8bR58)

 it doesn't seem to be called anywhere, and isn't implemented anywhere

miha-plesko commented:
Let me walk you through:
1 . user writes NodeJS application
2 . `npm install`
3 . `unik prepare-compiler --base osv --language java --provider openstack` -> manifest.yaml is created with following content (literally):
   ```yaml
   # Replace FILENAME with path to your node entrypoint filepath relative to the project root. E.g. server.js
   main_file: FILENAME
   ```
4 . user opens manifest.yaml and fills-in the parameters
5 . `unik build ...`
6 . `unik run ...`


PS: `Bootstrap()` methods were there by accident so I've removed them. I wanted to implement `unik prepare-compiler`with them, but then decided to ask you first.

I'm looking forward to hear your opinion.

ilackarms commented:
we should discuss that as a separate PR ;)

miha-plesko commented:
Good point. Let's focus on changes here first.

Would you argue with the refactoring that I proposed here for the OSv compiler? I always found myself confused when I've seen osv-java.go and osv-[vmware|virtuabox|aws].go in the same folder. So I wanted so separate them as clearly as possible to end the confusion once and for all, hence the refacotring :smile:

gberginc commented on pkg/compilers/includeos/includeos_qemu.go:
```diff
@@ -50,3 +50,7 @@ func (i *IncludeosQemuCompiler) findFirstImageFile(directory string) (string, er
 }
+
+func (r *IncludeosQemuCompiler) Usage() string {
+	return "TODO: usage for IncludeosQemuCompiler"
```
Write something like "No usage information available" or something like that. The TODO is for you or @ilackarms, not for the user interested in understanding how to use the compiler.

miha-plesko replied:
Done.

gberginc commented on cmd/describe-compiler.go:
```diff
@@ -0,0 +1,68 @@
+	Use:   "describe-compiler",
+	Short: "Describe compiler usage",
+	Long: `Describes compiler usage.
+You must provide triple base-language-provider to specify what compiler to describe.`,
```
I am still confused whether the `provider` is actually required here. Does any of the manifests rely on this info?

miha-plesko replied:
The main reason to use triple `base,lang,prov` instead of tuple `base,lang` is that in UniK, compiler **is** a triple. For example here: [link](https://github.com/emc-advanced-dev/unik/blob/master/pkg/compilers/names.go#L30). If I wanted to use tuple (and I did initially) then I would need to implement another mapping similar to the one in daemon: [link](https://github.com/emc-advanced-dev/unik/blob/master/pkg/daemon/daemon.go#L224-L379).

> Does any of the manifests rely on this info?
No. But I can think of a case where we might want to have it provider-dependent in the future. For example, with OSv we could make all compilers to support additional setting "enable_cloud_init" which would then automatically add cloud init Capstan package.

gberginc reviewed (commented):
I didnt do a complete review right now - apart from the comments, I wonder why you have to do all the changes with the bootstrapper thing? Can this be split into the functionality providing usage information and the other that goes towards the OSv/Capstan integration?

Commit:
Refactor OSv compiler

OSv compiler was misorganized, it was difficult to understand
the relation between osv-java.go and osv-virtualbox.go both sitting
in the same folder. Not any more - first refactored directory structure
of the pkg/compilers/osv to basically look like:
```
osv
 |- osvbootstrap
   |- virtualbox.go
   |- vmware.go
   |- aws.go
 |- java.go
```
Directly inside `osv` folder reside the actual compilers that make
use of Capstan (to compose package for selected language). Then there
is the `osvbootstrap` folder containing post-processing logic
(called "Bootstrapper") to carry out conversion of Capstan result into
appropriate image format (for the provider specified). The actual compiler
gets reference to the Bootstrapper and invokes its method
Bootstrap() to perform the conversion.

And that's it, code is more readable and managable.

Signed-off-by: Miha Pleško <miha.plesko@xlab.si>

Commit:
Add "unik describe-compiler" CLI

Currently user has to consult README to see how her
application must be configured (e.g. what config files
should look like) in order to run it with UniK.
Not any more! Introduced command
```
unik describe-compiler --base <> --language <> --provider <>
```
that prints usage information to the console.

Signed-off-by: Miha Pleško <miha.plesko@xlab.si>

miha-plesko reviewed (commented):
Thanks @gberginc , I've answered to your questions and repushed. Also, I've extracted the OSv compiler refactoring into antoher [PR](https://github.com/emc-advanced-dev/unik/pull/111) so I'm closing this one until the refactoring gets merged.

miha-plesko commented:
Closing this temporarily, until we merge OSv refactoring implemented in [this PR](https://github.com/emc-advanced-dev/unik/pull/111) .
